File: js/src/Checkout.js

class Checkout {
    constructor(rules) {
        this.basket = []
        this.rules = rules
        this.result = 0
        this.items2 = []
        this.hasDiscount = false
    }

    scan(item) {
        this.basket.push(item)
    }

    total() {
        this.basket.forEach(e => {   
            this.result += this.rules.table()[e]
        })

        this.result -= this.discountForA(this.basket) 
        this.result -= this.discountForB(this.basket)
        this.result -= this.discountForABC(this.basket)
        
        return this.result
    }

    discountForA(items){
        const discounts = Math.floor(items.join('').match(/A/g).length/3)
        const value = discounts * 20
        if(discounts > 0){
            this.hasDiscount = true   
        }
        return value
    }

    discountForB(items){
        const discounts =  Math.floor(items.join('').match(/B/g).length/2)
        const value = discounts * 15
        if(discounts > 0){
            this.hasDiscount = true
        }
        return value
    }

    discountForABC(items) {
        if (!this.hasDiscount && (items.includes('A') && items.includes('B') && items.includes('C'))) {
            return 5
        } else {
            return 0
        }
    }
}

module.exports = Checkout

File: ruby/lib/checkout.rb
class Checkout
  def initialize(rules)
    @items = []
    @rules = rules
    @result = 0
    @items2 = []
  end

  def scan(item)
    @items << item
  end

  def total
    @items.each do |e|
      @items2 << e
      @result += @rules.table[e]
      i = 0
      @items2.each { |ee| i += 1 if ee == e }
      e == 'A' && i % 3 == 0 ? @result -= 20 : e == 'B' && i % 2 == 0 ? @result -= 15 : @result
    end
    @result
  end
end
